### PR TITLE
PAYARA-4158 Allow create-domain command to configure the Hazelcast Configuration 

### PR DESCRIPTION
**Pull request by MeroRai**
# Description
This is a feature to allow `create-domain` command to configure the Hazelcast Configuration. This makes it possible to configure the Hazelcast Configuration before starting the domain.
 
**Following new command options have been added to the `create-domain` command:**

* **`--hazelcastdasport`** The port to run Hazelcast on for the DAS. If this port is busy, the port specified will be incremented until a valid port is found. The `--hazelcastdasport` option cannot be used with the --portbase option. The default value is 4900. The `--hazelcastdasport` option overrides the `hazelcast.das.port` property of the `--domainproperties` option.

* **`--hazelcaststartport`** The port the other Payara Server instances use to run Hazelcast on. If this port is busy, the port specified will be incremented until a valid port is found. The `--hazelcaststartport` option cannot be used with the `--portbase` option. The default value is 5900. The `--hazelcaststartport` option overrides the `hazelcast.start.port` property of the `--domainproperties` option.

**Following domain properties have been added to the `create-domain` command:**
* **`--hazelcastautoincrement`** By default the cluster uses the next unoccupied port that is available starting with the start port. When auto-increment is turned off an occupied start port results in a startup failure instead. The `--hazelcastautoincrement` option overrides the hazelcast.auto.increment property of the `--domainproperties` option. 

* **` hazelcast.das.port`** This property specifies the port number of the port that is used to run. Hazelcast on for the DAS. If this port is busy, the port specified will be incremented until a valid port is found.  Valid values are 1-65535. On UNIX, creating sockets that listen on ports 1-1024 requires superuser privileges. The hazelcast.das.port property is overridden by the `--hazelcastdasport` option.

* **`hazelcast.start.port`** This property specifies the port number of the port the other Payara Server instances use to run Hazelcast on. If this port is busy, the port specified will be incremented until a valid port is found. Valid values are 1-65535. On UNIX, creating sockets that listen on ports 1-1024 requires superuser privileges. The hazelcast.start.port property is overridden by the  `--hazelcaststartport` option.

* **`hazelcast.auto.increment`** This property specifies whether or not to use the next unoccupied port that is available starting with the start port. When auto-increment is turned off an occupied start port results in a startup failure instead. The `hazelcast.auto.increment` property is overridden by the `--hazelcastautoincrement` option.

It also shifts the Hazelcast DAS and the Start port according to the value of `--portbase` command option, similar to the other port values.

# Testing

### New tests
None

### Testing Performed
* Running the `create-domain` command to test the functionality of the newly implemented command options. E.g. running `asadmin create-domain --hazelcastdasport 7900 --hazelcaststartport 8900 --hazelcastautoincrement true testDomain` command. Gave following output:

```
Using default port 4848 for Admin.
Using default port 8080 for HTTP Instance.
Using default port 7676 for JMS.
Using default port 3700 for IIOP.
Using default port 8181 for HTTP_SSL.
Using default port 3820 for IIOP_SSL.
Using default port 3920 for IIOP_MUTUALAUTH.
Using default port 8686 for JMX_ADMIN.
Using default port 6666 for OSGI_SHELL.
Using default port 9009 for JAVA_DEBUGGER.
Using port 7900 for Hazelcast DAS.
Using port 8900 for Hazelcast Start.
Distinguished Name of the self-signed X.509 Server Certificate is:
[CN=dev,OU=Payara,O=Payara Foundation,L=Great Malvern,ST=Worcestershire,C=UK]
Distinguished Name of the self-signed X.509 Server Certificate is:
[CN=dev-instance,OU=Payara,O=Payara Foundation,L=Great Malvern,ST=Worcestershire,C=UK]
Domain testDomain created.
Domain testDomain admin port is 4848.
Domain testDomain allows admin login as user "admin" with no password.
Command create-domain executed successfully.
```
* Or running `create-domain --domainproperties hazelcast.das.port=7900:hazelcast.auto.increment=false:hazelcast.start.port=8900 testDomain` command. Gave following output:

```
Using default port 8080 for HTTP Instance.
Using default port 7676 for JMS.
Using default port 3700 for IIOP.
Using default port 8181 for HTTP_SSL.
Using default port 3820 for IIOP_SSL.
Using default port 3920 for IIOP_MUTUALAUTH.
Using default port 8686 for JMX_ADMIN.
Using default port 6666 for OSGI_SHELL.
Using default port 9009 for JAVA_DEBUGGER.
Using port 7900 for Hazelcast DAS.
Using port 8900 for Hazelcast Start.
Distinguished Name of the self-signed X.509 Server Certificate is:
[CN=dev,OU=Payara,O=Payara Foundation,L=Great Malvern,ST=Worcestershire,C=UK]
Distinguished Name of the self-signed X.509 Server Certificate is:
[CN=dev-instance,OU=Payara,O=Payara Foundation,L=Great Malvern,ST=Worcestershire,C=UK]
Domain testDomain created.
Domain testDomain admin port is 4848.
Domain testDomain allows admin login as user "admin" with no password.
Command create-domain executed successfully.
```
* Local Build Test

### Testing Environment
Zulu JDK 1.8_222 on Elementary OS 0.4.1 Loki with Maven 3.5.4

# Documentation
TODO

### CODE REVIEW

File: nucleus/admin/server-mgmt/src/main/java/com/sun/enterprise/admin/servermgmt/domain/DomainBuilder.java [hunk 1]
```diff
@@ -96,7 +95,7 @@
     private final DomainConfig domainConfig;
     private JarFile templateJar;
     private DomainTemplate domainTemplate;
-    private final Properties defaultPortValues = new Properties();
+    private final Properties defaultProertiesValue = new Properties();
```
Comment (jGauravGupta) [inline nucleus/admin/server-mgmt/src/main/java/com/sun/enterprise/admin/servermgmt/domain/DomainBuilder.java]: typo in `defaultPropertiesValue ` variable

File: appserver/admin/gf_template_web/src/main/resources/config/domain.xml [hunk 2]
```diff
@@ -395,7 +396,7 @@
              </message-security-config>
          </security-service>
          <transaction-service tx-log-dir="${com.sun.aas.instanceRoot}/logs" automatic-recovery="true" />
-         <hazelcast-runtime-configuration></hazelcast-runtime-configuration>
+            <hazelcast-config-specific-configuration></hazelcast-config-specific-configuration>
```
Comment (jGauravGupta) [inline appserver/admin/gf_template_web/src/main/resources/config/domain.xml]: Why `hazelcast-runtime-configuration` element renamed to `hazelcast-config-specific-configuration` ?

Comment (MeroRai) [reply appserver/admin/gf_template_web/src/main/resources/config/domain.xml]: The `hazelcast-runtime-configuration` element is used to configure Domain wide Hazelcast Settings, and `hazelcast-config-specific-configuration` is only used to configure specific config.

**Comment by MeroRai** [comment]
jenkins test please

File: nucleus/admin/server-mgmt/src/main/manpages/com/sun/enterprise/admin/servermgmt/cli/create-domain.1 [hunk 3]
```diff
@@ -106,6 +109,10 @@ OPTIONS
 
            *   Felix shell service port for OSGi module management: portbase +
                66
+           
+           *   Hazelcast DAS port: portbase + 0
+           
+           *   Hazelcast Start port: portbase + 0
```
Comment (jGauravGupta) [inline nucleus/admin/server-mgmt/src/main/manpages/com/sun/enterprise/admin/servermgmt/cli/create-domain.1]: Suffix typo here, It should be 49 & 59.

**Pull request merged**